clang/lib/AST/ItaniumMangle.cpp

```diff
@@ -7326,11 +7326,13 @@ void ItaniumMangleContextImpl::mangleCXXCtorVTable(const CXXRecordDecl *RD,
                                                    raw_ostream &Out) {
   // <special-name> ::= TC <type> <offset number> _ <base type>
   CXXNameMangler Mangler(*this, Out);
+  QualType RDType = getASTContext().getRecordType(RD);
+  QualType TypeType = getASTContext().getRecordType(Type);
   Mangler.getStream() << "_ZTC";
-  Mangler.mangleNameOrStandardSubstitution(RD);
+  Mangler.mangleType(RDType);
```

Comment:
The only difference here is whether we perform / add a substitution for RD.

We only ever see a problem for `mangleCXXCtorVTable` because it's the only user of `mangleNameOrStandardSubstitution` which mangles anything at all after a call to that function.

I can't find this production in the ItaniumABI docs, but if the comment is right, I think performing this substitution is correct.

I'd prefer we fix `mangleNameOrStandardSubstitution` itself.
We could rename it to `mangleCXXRecordDecl`, have it take a CXXRecordDecl*, and handle the substitutions there.
This would avoid the same trap in a potential future user.

CC @rjmccall @zygoloid

Reply:
> I can't find this production in the ItaniumABI docs, but if the comment is right, I think performing this substitution is correct.

I also can't find any Itanium ABI documentation, but I found that GCC follows the same process. So, I believe the comment is correct, and it is necessary to perform the substitution for RD.

Reply:
> I'd prefer we fix `mangleNameOrStandardSubstitution` itself.
> We could rename it to `mangleCXXRecordDecl`, have it take a CXXRecordDecl*, and handle the substitutions there.
> This would avoid the same trap in a potential future user.

After rechecking the code, I found that the users of the `mangleNameOrStandardSubstitution` function are `mangleCXXCtorVTable`, `mangleCXXVTable`, and `mangleCXXVTT`. In the Itanium ABI documentation, these situations are considered as 'Type'. So, maybe we can remove `mangleNameOrStandardSubstitution` and then use `mangleType`.

Reply:
Yes, although as I said before, there should be no observable change for these other users, as we don't perform other manglings either before or after calling `mangleNameOrStandardSubstitution`.

I would prefer the ergonomics of a function which takes a CXXRecordDecl. MangleType is also more complex and  will perform a bunch of checks which are not necessary if you only would ever have a RecordType, so there is also some aspect of micro-optimization and ease of debugging in having a specialized function.

Reply:
Thanks, i see what you mean.